channeldb/migration29/iterator.go

```diff
@@ -0,0 +1,257 @@
+	// another level, for this channel itself.
+	chanBucket := chainBucket.NestedReadWriteBucket(ul.fundingOutpoint)
+	if chanBucket == nil {
+		return nil, mig25.ErrChannelNotFound
+	}
+
+	return chanBucket, nil
+}
+
+// findNextMigrateHeight finds the next commit height that's not migrated. It
+// returns the commit height bytes found. A nil return value means the
+// migration has been completed for this particular channel bucket.
+func findNextMigrateHeight(chanBucket kvdb.RBucket) []byte {
+	// Read the old log bucket. The old bucket doesn't exist, indicating
+	// either we don't have any old logs for this channel, or the migration
+	// has been finished and the old bucket has been deleted.
+	oldBucket := chanBucket.NestedReadBucket(
+		revocationLogBucketDeprecated,
+	)
+	if oldBucket == nil {
+		return nil
+	}
+
+	// Acquire a read cursor for the old bucket.
+	oldCursor := oldBucket.ReadCursor()
+
+	// Read the new log bucket. The sub-bucket hasn't been created yet,
+	// indicating we haven't migrated any logs under this channel. In this
+	// case, we'll return the first commit height found from the old
+	// revocation log bucket as the next height.
+	logBucket := chanBucket.NestedReadBucket(revocationLogBucket)
+	if logBucket == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Acquire a read cursor for the new bucket.
+	cursor := logBucket.ReadCursor()
+
+	// Read the last migrated record. If the key is nil, we haven't
+	// migrated any logs yet. In this case we return the first commit
+	// height found from the old revocation log bucket.
+	migratedHeight, _ := cursor.Last()
+	if migratedHeight == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Read the last height from the old log bucket. If the height of the
+	// last old revocation equals to the migrated height, we've done
+	// migrating for this channel.
+	endHeight, _ := oldCursor.Last()
+	if bytes.Equal(migratedHeight, endHeight) {
+		return nil
+	}
+
+	// Now point the cursor to the migratedHeight. If we cannot find this
+	// key from the old log bucket, the database might be corrupted. In
+	// this case, we would return the first key so that we would redo the
+	// migration for this chan bucket.
+	matchedHeight, _ := oldCursor.Seek(migratedHeight)
+	if matchedHeight == nil {
+		nextHeight, _ := oldCursor.First()
+		return nextHeight
+	}
+
+	// Otherwise, find the next height to be migrated.
+	nextHeight, _ := oldCursor.Next()
+
+	return nextHeight
+}
+
+// locateNextUpdateNum returns a locator that's used to start our migration. A
+// nil locator means the migration has been finished.
+func locateNextUpdateNum(openChanBucket kvdb.RwBucket) (*updateLocator, error) {
+	locator := &updateLocator{}
+
+	// cb is the callback function to be used when iterating the buckets.
+	cb := func(chanBucket kvdb.RwBucket, l *updateLocator) error {
+		locator = l
+		updateNum := findNextMigrateHeight(chanBucket)
+
```

Comment:
I think we should log the height here, so we can use it to print some more progress during the migration. Otherwise, users may just see it sitting there for minutes and wonder what's going on.

Reply:
Added more verbose logs and debug logs.